src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Workbook.kt

```diff
@@ -2,5 +2,7 @@ package org.wycliffeassociates.otter.common.data.workbook
 
 data class Workbook(
     val source: Book,
-    val target: Book
+    val target: Book,
+    val sourceLanguageSlug: String?,
+    val targetLanguageSlug: String?
```

Comment:
Why are these nullable?

Reply:
See line 49 of WorkbookRepository.kt. Collection's `resourceContainer` field is nullable, which is where we get the language slug. Should WorkbookRepository.get() just fail (throw an exception??) if we can't get the language slug due to a null `resourceContainer`?

Reply:
Null values are like checked exceptions in Java (and this is a good thing if handled right). If handled correctly at the first point where nulls could get introduced, then you don't have to worry about null values creeping into later parts of the app. If you do let them in (and kotlin makes it so you the programmer explicitly let null values enter your code, the language doesn't by default), at each point further down the chain you have to handle the null, and most of the time you're not going to know what to do because it's this "...how could something get this far along if it doesn't?" The null most of the time needs to be handled at the point it gets introduced, and from that point on, things shouldn't be nullable. There are some exceptions, where null values are used as a kind of "optional" type. But as a result, the "option" needs to be checked in all places it is considered.

Now; from a correctness standpoint, a translation project has a source (it has to, otherwise we have no idea any information like how many chapters, chunks, etc) and sources have to have resource contianers. We legitimately CANNOT move forward without a source. So a workbook can't be created without one. A workbook cannot be created without a target language either, that doesn't make sense.

So, if we have to have a source, and we have to have a target, those works both have to have languages.

Allowing them to be nulls is therefore by definition a bug (may not present itself, but why even allow for it?). 

What has a null resource container?

Reply:
Very good points. 

I think no Collection should have a null RC. From inspecting the code again, we create Collections with a null resource container as an intermediate step during the import process, but immediately after they are mapped to CollectionEntities, we set the dublinCoreFk to what it should be before saving the entity to the database. (dublinCoreFk is non-nullable in the DB.) So, we really shouldn't be encountering a Collection with a null resource container that came from the database- if we encounter that, we should handle it immediately.